protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+
+- Peer Routing - Mechanism to find a Peer in a network. This Routing can be done recursively, iteratively or even in a broadcast/multicast mode.
+- Swarm - Handles everything that touches the 'opening a stream' part of libp2p, from protocol muxing, stream muxing, NAT Traversal, Connection Relaying, while being multitransport
+- Distributed Record Store - A system to store provider records and relay records
+
+Each of these subsystem exposes a well known interface (see chapter 6 for Interfaces) and may use eachother in order to fulfil their goal. A global overview of the system is:
+
+```
+┌─────────────────────────────────────────────────────────────────────────────────┐
+│                                  libp2p                                         │
+└─────────────────────────────────────────────────────────────────────────────────┘
+┌─────────────────────────┐┌──────────────────────────┐┌──────────────────────────┐
+│       Peer Routing      ││      Swarm               ││ Distributed Record Store │
+└─────────────────────────┘└──────────────────────────┘└──────────────────────────┘
+```
+
+## 4.1 Peer Routing
+
+A Peer Routing subsystem, exposes an interface to identify which peers should a message be routed in the DHT. It receives a key and must return one or more PeerInfo objects.
+
+We present two examples of possible Peer Routing subsystems, the first based on a the Kademlia DHT and the second based on mDNS. Nevertheless, other Peer Routing mechanisms might be implemented, as long as they fulfil the same expectation and interface.
+
+### 4.1.1 kad-routing
+
+kad-routing implements the Kademlia Routing table, where each peer holds a set of k-buckets, each of them containing several PeerInfo from other peers in the network. In order to find the whereabouts of these peers, it implements 3 discovery mechanisms:
+
+- mDNS-discovery
+- random-walk
+- bootstrap-list
+
+```
+┌────────────────────────────────────────────────────────────────┐
+│       Peer Routing                                             │
+│┌──────────────────────────────────────────────────────────────┐│
+││ kad-routing ┌──────────────┐┌──────────────┐┌──────────────┐ ││
+││             │mDNS-discovery││random-walk   ││bootstrap-list│ ││
+││             └──────────────┘└──────────────┘└──────────────┘ ││
+│└──────────────────────────────────────────────────────────────┘│
+│┌──────────────────────────────────────────────────────────────┐│
+││ mDNS-routing                                                 ││
+│└──────────────────────────────────────────────────────────────┘│
+│┌──────────────────────────────────────────────────────────────┐│
+││ other-routing-mechanisms                                     ││
+│└──────────────────────────────────────────────────────────────┘│
+└────────────────────────────────────────────────────────────────┘
+```
```

Comment:
this diagram is misleading, vertical placement tends to imply "things rely on the things below them" where here i think you want to show how protocols "peer" with each other. useful to represent them horizontally in that case.

(i know horizontal is challenged, but most networks people who read a vertical stack of protocols take that to mean well-layered dependence relations, when here that isn't the case)